docs/architecture/adr-110-remote-mempool.md

```diff
@@ -0,0 +1,364 @@
+
+Additionally, when transactions are received by the local mempool instance from
+foreign mempool instances, it is expected that the local mempool instance will
+send the transaction to the application for validation (similar to how `CheckTx`
+currently functions).
+
+#### Scalable mempool
+
+The recommended high-level architecture for a modular Comet validator with an
+externalized, scalable mempool, is as follows.
+
+```mermaid
+flowchart LR
+    comet[Comet Validator]
+    app[Application]
+    mempoolrouter[Mempool Router]
+    mempool1[Mempool Instance 1]
+    mempool2[Mempool Instance 2]
+    mempoolN[Mempool Instance N]
+    user[User]
+
+    comet --> app
+
+    comet -- 4. Reap transactions --> mempoolrouter
+    user -- 1. Submit transactions --> app
+    app -- 2. CheckTx --> app
+    app -- 3. Publish transactions --> mempoolrouter
+
+    mempoolrouter --> mempool1
+    mempoolrouter --> mempool2
+    mempoolrouter --> mempoolN
+
+    mempool1 -- CheckTx --> app
+    mempool2 -- CheckTx --> app
+    mempoolN -- CheckTx --> app
+```
+
+Here the interaction is the same as the simple variant discussed earlier, but
+instead of interacting directly with the mempool, all interactions are with some
+form of routing mechanism. For example, this could be a reverse proxy like nginx
+configured to deliver requests to mempool instances in a round-robin fashion.
+
+This configuration would be application-specific, and would need to be set up
+correctly by the operator for the specific application.
+
+### Configuration
+
+The following change to the `config.toml` file is envisaged:
+
+```toml
+[mempool]
+# The type of mempool for this CometBFT node to use.
+#
+# Valid types of mempools supported by CometBFT:
+# - ""       : Default clist mempool with flooding gossip protocol
+# - "remote" : Remote mempool in a separate process
+type = "remote"
+
+#
+# Configuration specific to the remote mempool. If mempool.type is not "remote",
+# this section will be ignored.
+#
+# A remote mempool is only usable by a validator node. Turning on the remote
+# mempool for a full node will simply disable any mempool-related functionality
+# on that full node, and the full node will not interact with any mempool at
+# all.
+#
+[mempool.remote]
+# The base URL for the gRPC interface to the remote mempool.
+url = "http://localhost:28880"
+
+# The timeout for fetching transaction data during PrepareProposal. Exceeding
+# this timeout will result in a panic.
+timeout_fetch = "5s"
+
+# The timeout for reaping (removing) transaction data after a block has been
+# committed. Exceeding this timeout will result in a panic.
+timeout_reap = "1s"
+
+# The timeout for initiating the TxsAvailable call.
+timeout_txs_available = "1s"
+
+# The maximum number of times to retry calling the TxsAvailable streaming
+# endpoint if it fails. Exceeding this number of retries (e.g. if the connection
+# to the remote mempool is down) will result in a panic.
+txs_available_retries = 3
+```
```

Comment:
I second @melekes's opinion or at least I would like to understand why exactly is it needed to panic. Expecially on the first try for getting transactions. Do we panic now when the mempool is empty and we can produce empty blocks?